apps/server/src/shared/domain/rules/base.rule.ts

```diff
@@ -0,0 +1,171 @@
+import { Collection } from '@mikro-orm/core';
+import { UnauthorizedException } from '@nestjs/common';
+import { Role } from '../entity/role.entity';
+import { User } from '../entity/user.entity';
+import { IEntity, IEntityWithSchool } from '../interface';
+import { Actions } from './actions.enum';
+
+export abstract class BaseRule {
+	abstract hasPermission(user: User, entity: IEntity, action: Actions): boolean;
+	/**
+	 * Recursively resolve all roles and permissions of a user.
+	 * IMPORTANT: The role collections of the user and nested roles will not be loaded lazily.
+	 * Please make sure you populate them before calling this method.
+	 * @param user
+	 * @returns
+	 */
+	resolvePermissions(user: User): string[] {
+		if (!user.roles.isInitialized(true)) {
+			throw new Error('Roles items are not loaded.');
+		}
+		const rolesAndPermissions = this.resolvePermissionsByRoles(user.roles.getItems());
+
+		return rolesAndPermissions;
+	}
+
+	private resolvePermissionsByRoles(inputRoles: Role[]): string[] {
+		let permissions: string[] = [];
+
+		for (let i = 0; i < inputRoles.length; i += 1) {
+			const role = inputRoles[i];
+			if (!role.roles.isInitialized(true)) {
+				throw new Error('Roles items are not loaded.');
+			}
+			const innerRoles = role.roles.getItems();
+			permissions = [...permissions, ...role.permissions];
+
+			if (innerRoles.length > 0) {
+				const subPermissions = this.resolvePermissionsByRoles(innerRoles);
+				permissions = [...permissions, ...subPermissions];
+			}
+		}
+
+		permissions = [...new Set(permissions)];
+
+		return permissions;
+	}
+
+	/**
+	 * Determines whether a user has all the rights specified in requiredPermissions
```

Comment:
The descriptions for hasAllPermssions,  checkAllPermissions, hasOneOfPermissions do not include any information that is not already showed by the method name and the props. But increase the file lengh and decrease the readability of the file it self a lot.

Removing of  @param XXX is the minimum. But only hasRole, hasAccessToEntity, hasPermission include comments that give more hints about the method that i can confirm that we let stand a comment for it.

Reply:
I dont't really see the point against JSDoc. Why should we remove it? I think, it's really useful.